source/opt/loop_descriptor.cpp

```diff
@@ -639,7 +691,17 @@ int64_t Loop::GetIterations(SpvOp condition, int64_t condition_value,
   return result;
 }
 
-ir::Instruction* Loop::FindInductionVariable(
+// Returns the list of induction variables within the loop.
+void Loop::GetInductionVariables(
+    std::vector<ir::Instruction*>& induction_variables) const {
+  for (ir::Instruction& inst : *loop_header_) {
+    if (inst.opcode() == SpvOp::SpvOpPhi) {
+      induction_variables.push_back(&inst);
```

Comment:
Hmm, is this really correct?  There may be Phi instructions for variables that are not really induction variables.  You also need the Phi to have a cycle from within the loop.  Could you point at where in the literature you are taking induction variable detection from?  Having a paper/article source makes documentation much easier.

This is a contrived example, but |z| below is not really an induction variable:

```
#version 450
layout(location  = 0) in flat int  y;
layout(location  = 0) out int  outparm;

void main()
{
  int i, z=3;
  #pragma unroll
  for (i = 0; i < 16; i++) {
    if (y > 10) z+=5;
   }
  outparm = z;
}
```

Reply:
The reason we need this is that we need to track all the phis which are coming into the loop so we can change the uses in the unrolled bodies and clean up uses outside the loop after fully unrolling. So we still need to track them whether or not they are strictly induction variables. It's a generic approach which we need for phis so we aren't following any paper on it, I think it should probably be renamed to header phis or something. 

With the work we are doing on dependence analysis we have a much better solution for finding induction variables and eliminating a lot of existing code by using the newer approach. (using the scalar analysis graph, this is closer to my understanding from literature). It will be useful to have a generic induction variable detection solution, and it lets us clean up a lot of code, but it would be good if it could be deferred to a later patch once we have some more of the analysis tools coming soon.

Reply:
Sure.  This test will currently return many false positives, but they're harmless.